Contact overview page for admins

basilesimon opened the pull request:
e.g. [http://localhost:3000/contacts/0adfc2c8-c35f-4ad3-8b4a-3118a74f7797](http://localhost:3000/contacts/0adfc2c8-c35f-4ad3-8b4a-3118a74f7797)

based on [🎨 single contact designs](https://design.penpot.app/#/workspace/887044a0-9c39-11eb-b4cd-81aaa3a367f9/3342c4a0-46c8-11ec-8508-6194167bb409?page-id=224b7680-576f-11ec-bc54-4148703305a8)

--

Todo list:
- [x] Clarify what "roles" are: an array of dates?
- [x] Implement "Security" section
- [x] Include default "-" when no option: out of this PR's scope
- [x] Localise [this para](https://github.com/beabee-communityrm/beabee-frontend/pull/104/files#diff-9ca9ac60d0dd5675df8fef4e5a2c075630d5c7659c63a9f501e30f6adcdbf54aR68-R72)
- [x] The "About" form should support updating the contact/member: `updateMember`
- [x] Switch list redirect from `/members/:id` to `/contacts/:id`
- [x] Re-implement grid layout to make sections close to each other vertically
- [x] Add an `AppTextArea` component?

basilesimon committed:
feat: stubs for additional sections

basilesimon committed:
chore: update locales

basilesimon committed:
feat: replace placeholders with API-driven values

basilesimon committed:
feat: scaffolding for admin notes about a contact

basilesimon committed:
refactor: use AppButton, AppInput

basilesimon committed:
feat: support (and implement) a form textarea

basilesimon committed:
style: rejig CSS Grid for more natural flow

basilesimon committed:
fix: link to /contacts, not /members

basilesimon commented:
"Include default "-" when no option"

@wpf500 what do you recommend in this instance?
I could probably `compute()` UI-friendly strings, instead of relying on `contact`'s raw data fields.

basilesimon commented:
re: https://github.com/beabee-communityrm/beabee-frontend/pull/104/commits/0ff598f9579a42652b120e60e5e9f375160976ff

how do I stop the whole app redrawing? you mentioned a magic component... 😉

basilesimon committed:
feat: localise annotation copy

basilesimon committed:
feat: form handles editing of notes and description

wpf500 committed:
feat: add with parameter to fetchMember for new subresources

wpf500 commented:
> "Include default "-" when no option"
> 
> @wpf500 what do you recommend in this instance? I could probably `compute()` UI-friendly strings, instead of relying on `contact`'s raw data fields.

I think for this I would opt to handle it in the `AppInfoListItem` component, so if it gets an empty value it defaults to "-"

basilesimon committed:
feat: soft, router-link replaces full-page reload

basilesimon committed:
feat: include currency in contribution amount display

basilesimon committed:
feat: include contribution details

basilesimon committed:
feat: implement roles displaying – refactor ListItem, now takes a <slot>

basilesimon committed:
fix: v-for one level higher, simplifies DOM tree

basilesimon committed:
feat: localise roles, and their duration

basilesimon committed:
feat: mock up security section, password reset and login links

basilesimon commented:
@aiscarvalho - @aiscarvalho 

There are several design changes that came up during the initial implementation of this page. Here's the overview:

![image](https://user-images.githubusercontent.com/5701152/155364740-293f7360-a0a5-48ef-a86b-c1365fa1f4d3.png)

### On Roles

![image](https://user-images.githubusercontent.com/5701152/155368556-8dc3cab8-24c9-4804-9ac8-c5719cf2f416.png)

- I've added a unicode arrow in lieu of the dash separating the dates. Hope that's fine!
- The little green badge is using our "success" colour if the role is current, "warning" otherwise. As intended?
- For current roles (see "superadmin" above), what do we want to display? I added "onwards" but that doesn't feel like a good candidate for localisation.

### On Security

https://user-images.githubusercontent.com/5701152/155365510-486dfc0c-caa0-43b1-a5a9-a1b63f6989d6.mov

- Will and I changed the layout slightly, it felt a bit more natural. Could you review?
- Clicking one of the buttons disables both and displays the result in a read-only input field below. How's that?

wpf500 commented on src/modules/contacts/pages/ContactOverviewPage.vue:
```diff
@@ -17,34 +17,196 @@
+        />
+        <AppInfoListItem
+          :name="t('contacts.data.deliveryOptIn')"
+          :value="contact.profile.deliveryOptIn"
```
This should be internationalised using `common.yes` and `common.no`

basilesimon replied:
Done: https://github.com/beabee-communityrm/beabee-frontend/pull/104/commits/578f343a4e4679ed35823e36a989aa613b9789b8

wpf500 commented on src/modules/contacts/pages/ContactOverviewPage.vue:
```diff
@@ -17,34 +17,196 @@
+        <AppInfoListItem
+          v-if="contact.contribution.type === ContributionType.GoCardless"
+          :name="t('contacts.data.payingFee')"
+          :value="contact.contribution.payFee"
```
This should be internationalised using common.yes and common.no

basilesimon replied:
Done: https://github.com/beabee-communityrm/beabee-frontend/pull/104/commits/578f343a4e4679ed35823e36a989aa613b9789b8

wpf500 commented on src/modules/contacts/pages/ContactOverviewPage.vue:
```diff
@@ -17,34 +17,196 @@
+      <AppInfoList v-for="role in contact.roles" :key="role.role">
+        <AppInfoListItem :name="t(`common.role.${role.role}`)">
+          <!-- @TODO: make component -->
+          <span class="inline-block w-2 h-2 bg-success rounded-full"></span>
```
Make this into a little component which takes a colour as an attribute

basilesimon replied:
Done: https://github.com/beabee-communityrm/beabee-frontend/pull/104/commits/5e4f54269287cd70ffe74ac20a25b3c328a4b585

wpf500 commented on src/modules/contacts/pages/ContactOverviewPage.vue:
```diff
@@ -17,34 +17,196 @@
+        <AppButton
+          type="submit"
+          variant="link"
+          :disabled="!securityButtonsEnabled"
```
I'm not a huge fan of inverse logic where possible, could this be more readable as `securityButtonsDisabled` instead?

basilesimon replied:
Done: https://github.com/beabee-communityrm/beabee-frontend/pull/104/commits/578f343a4e4679ed35823e36a989aa613b9789b8

wpf500 commented on src/components/forms/AppTextArea.vue:
```diff
@@ -0,0 +1,68 @@
+      rounded
+      focus:outline-none focus:shadow-input
+    "
+    :type="inputType"
```
I don't think we need a type attribute for this component as it's not relevant to a `<textarea>`

basilesimon replied:
Done: https://github.com/beabee-communityrm/beabee-frontend/pull/104/commits/578f343a4e4679ed35823e36a989aa613b9789b8

wpf500 commented on src/components/forms/AppTextArea.vue:
```diff
@@ -0,0 +1,68 @@
+<template>
+  <label v-if="label" class="block mb-1.5 font-semibold" :for="inputType"
```
Hmm this isn't related to your task but I'm not sure how the `<label for="inputType">` and `<input id="inputType">` ever worked here, this clearly won't generate unique IDs! I'm guessing this means all our labels are broken if you try to use them to focus an input. I'll add an issue to fix this.

basilesimon replied:
Done: https://github.com/beabee-communityrm/beabee-frontend/pull/104/commits/578f343a4e4679ed35823e36a989aa613b9789b8

wpf500 commented on src/modules/contacts/pages/ContactOverviewPage.vue:
```diff
@@ -17,34 +17,196 @@
+        />
+        <AppInfoListItem
+          :name="t('contacts.data.groups')"
+          :value="contact.profile.newsletterGroups.join(',')"
```
Join with `', '` (a space after the comma) for slightly nicer formatting?

basilesimon replied:
Done: https://github.com/beabee-communityrm/beabee-frontend/pull/104/commits/578f343a4e4679ed35823e36a989aa613b9789b8

wpf500 reviewed (commented):
Looks great, nice work :+1:. A few comments here and there and some stuff left to do, but mostly because you are waiting on me to create the API calls ;)

basilesimon committed:
feat: make component from round badge, parameterise

basilesimon committed:
fix: code review

basilesimon commented:
> Looks great, nice work 👍. A few comments here and there and some stuff left to do, but mostly because you are waiting on me to create the API calls ;)

Thanks for the review, I've addressed all of your points.

aiscarvalho commented:
>     * I've added a unicode arrow in lieu of the dash separating the dates. Hope that's fine!

yes!

>     * The little green badge is using our "success" colour if the role is current, "warning" otherwise. As intended?

Apologies because we didn't document this in the design.
The coloured dot is related to the member role only, it's an indication of an active (=paying) or expired (=non-paying) member.
So now that you mentioned it we realise it can be read as related to the role duration.  Maybe it makes more sense to move it to the "Contribution" section instead of here. This way we don't need to have a visual element that only exists for this role.

Should we mock that change in the design?
 
>     * For current roles (see "superadmin" above), what do we want to display? I added "onwards" but that doesn't feel like a good candidate for localisation.

We'll get back to this tomorrow!
 
>     * Will and I changed the layout slightly, it felt a bit more natural. Could you review?

Looks great!

basilesimon committed:
fix: default value if empty field - close #106

wpf500 commented:
> Apologies because we didn't document this in the design.
> The coloured dot is related to the member role only, it's an indication of an active (=paying) or expired (=non-paying) member.
> So now that you mentioned it we realise it can be read as related to the role duration. Maybe it makes more sense to move it to the "Contribution" section instead of here. This way we don't need to have a visual element that only exists for this role.

I think the coloured dot should actually apply to the role for now. There is a lot of confusion between the member role and having an active contribution which we definitely need to revisit, but I don't think it should block this task. So for now all let's just keep this the same as in the legacy app: all roles have a status, status is:
```
green = dateAdded < now && (!dateExpires || dateExpires > now), red otherwise.
```

basilesimon committed:
fix: align list items on their baseline - close #105

rlafuente commented:
Picking up the open point about date ranges with no end date, using "onwards": 

Not sure whether there's a standard way to signal this, and personally I'd go for "2022/02/24 –⁠ present", but it's more of an editorial/copy decision that should take translation into account (good point @basilesimon). "Present" would probably be more translation-friendly, but it is understandable as it is and shouldn't constitute a blocker for this issue.

All good to go from our side :ok_hand:

basilesimon committed:
feat: improve wording

basilesimon commented:
> Not sure whether there's a standard way to signal this, and personally I'd go for "2022/02/24 –⁠ present", but it's more of an editorial/copy decision that should take translation into account

I love it – implemented it in commit above, thanks for your help.

basilesimon committed:
feat: logic to render coloured badge depending on role's currency

basilesimon commented:
@wpf500 I implemented your logic to decide which colour to render the badge in https://github.com/beabee-communityrm/beabee-frontend/pull/104/commits/5cc57adb2bcb9607193eebb2efbde05a0d2581e9 – can you review?

basilesimon committed:
chore: hide security section for now, waiting for API endpoints

basilesimon commented:
I re-targeted this PR, which will now land into `dev`

basilesimon committed:
Merge branch 'dev' into feat/contact-overview

aiscarvalho commented:
Looks good!

We've had a first view of the page and found the following items:
- in mobile, the overview list is mis-aligned with the others
- the save is not yet working: we wrote in the notes and description, saved and shifted to tab information; when we returned to overview notes and description were empty again
- the "save button" should be primary color instead of link
- the security section is not visible yet

![image](https://user-images.githubusercontent.com/122730/155574819-8321a639-c9b3-4741-89d0-04aea0f77699.png)

basilesimon commented:
Here's the fix I can offer you, given that email addresses are arbitrarily long:

![Screenshot 2022-03-01 at 10 27 36](https://user-images.githubusercontent.com/5701152/156142239-4130d789-2ef0-416d-b6f6-a21ae0cef96a.png)

![Screenshot 2022-03-01 at 10 27 52](https://user-images.githubusercontent.com/5701152/156142291-f205393a-fe66-46bd-9c44-2d9d11014096.png)

Will address **link colour** in a coming commit.

I left the **security section** hidden because we're missing the API side, so I'll turn this on in the future (you can toggle the CSS class in the meantime).

And about **saving the notes**, this is the meaning of my earlier message:

> The about section, which lets you annotate a contact, needs a bit of UX thinking, I believe. Empty textboxed make sense when there's no data yet, but what should we present when someone has notes written about them already?

aiscarvalho commented:
The alternative summary layout, for mobile, looks great! Very tidy and aligned now.
For ultra long email addresses we can break the line: the label first and the email below (if that makes sense to tackle).

Editorial: could we suggest using an em dash for empty fields?

The security section looks fine but the button style is off. These should be outlined buttons.

I had replied about **saving the notes** in Trello, so copying it here:
 
> thanks for bringing that up, it was something we [discussed over the design spec](https://trello.com/c/2htyLaCL/333-design-elements-contact-overview-page-%F0%9F%8E%A8).
>
> Our first take for the contact overview page was to show a comprehensive summary, and nothing was editable in this page but in its specific page. Will and Julia made some fair arguments over the need for quick note taking and so we kept the fields for notes, description and tags editable in this page. If there is information in these textboxes it should be editable right away, as it is in the legacy app (and the static spec :)

Almost ready to close!
